### PR TITLE
[ECO-4982] Integrate lifecycle manager into existing room operations

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
Replace the existing temporary implementations of room attach / detach / status with those provided by the room lifecycle manager.

Part of #47.

## Release Notes

- **New Features**
  - Enhanced room lifecycle management with new classes for managing room states and interactions.
  - Introduced asynchronous operations for attaching and detaching rooms, improving responsiveness.
  - Added a mock lifecycle manager for testing, allowing for better simulation of room behavior.
  - Implemented a new clock class to facilitate asynchronous sleep operations.
  - Updated `Room` protocol to support asynchronous status retrieval.

- **Bug Fixes**
  - Improved type handling in tests to align with updated room management structures.

- **Documentation**
  - Updated test methods to reflect changes in room lifecycle management and initialization parameters.

- **Chores**
  - Refactored existing tests to streamline lifecycle management verification and enhance clarity.

### CODE REVIEW

File: Tests/AblyChatTests/Mocks/MockRoomLifecycleManager.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        try attachResult.get()
+    }
+
+    func performDetachOperation() async throws {
+        detachCallCount += 1
+        guard let detachResult else {
+            fatalError("In order to call performAttachOperation, detachResult must be passed to the initializer")
+        }
+        try detachResult.get()
+    }
```
[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply Tests/AblyChatTests/Mocks/MockRoomLifecycleManager.swift:33]: Ah yes, will fix

**Comment by lawrence-forooghian** [comment]
To reviewers: please ignore the failure of the spec-coverage job; that's not related to this issue, will look at it separately.

**Review by umair-ably — APPROVED** [review]
nice work on the tests

File: Tests/AblyChatTests/DefaultRoomTests.swift [hunk 2]
```diff
@@ -11,187 +11,155 @@ struct DefaultRoomTests {
...
-        // Then: `attach(_:)` is called on each of the channels, the room `attach` call succeeds, and the room transitions to ATTACHED
-        for channel in channelsList {
-            #expect(channel.attachCallCounter.isNonZero)
+        // When: `attach()` is called on the room
+        let result = await Result { () async throws(ARTErrorInfo) in
+            do {
+                try await room.attach()
+            } catch {
+                // swiftlint:disable:next force_cast
+                throw error as! ARTErrorInfo
```
Comment (maratal) [inline Tests/AblyChatTests/DefaultRoomTests.swift:50]: Doesn't room already throws ARTErrorInfo? Why just re-throw it?

**Review by maratal — COMMENTED** [review]
I'm curious why rooms can't be their own lifecycle managers? Why separated object for just that?

**Comment by lawrence-forooghian** [comment]
> I'm curious why rooms can't be their own lifecycle managers? Why separated object for just that?

It's a good question; I think that I just did it because I’d been hearing so much about the JS SDK’s “room lifecycle manager” that when I came to implement it in Swift I just created a class of the same name without thinking much about it. We could revisit in the future, but I don't really feel like trying to do that now.

File: Tests/AblyChatTests/DefaultRoomTests.swift (hunk 2 above)
Comment (lawrence-forooghian) [reply Tests/AblyChatTests/DefaultRoomTests.swift:50]: Yes, it throws `ARTErrorInfo` but we don't currently communicate this in the type system (see #43), hence this cast.

Comment (lawrence-forooghian) [reply Tests/AblyChatTests/DefaultRoomTests.swift:50]: @umair-ably asked me to resolve this one so that we could merge it

**Pull request merged**